### DISCUSSION TITLE
Confused about RFID Player Control and Second Swipe

### DISCUSSION THREAD

**Discussion by arjunsarin** [MiczFlor/RPi-Jukebox-RFID#2117] [Q&A] [answered]
Hello!

I am struggling to get the desired player behavior when swiping RFID cards which start music playback. **My goal is, that placing the card starts player, removal stops it. Further, I want that the player resumes in that folder when the same card is placed again.** That's how the TonieBox is implementing the control and I like that approach.

I tried different settings like the "General RFID player control" in the web interface. This is bringing a nearly perfect behavior, however the playlist starts from the beginning when the card is placed again. How can I change this?

I tried several other settings like placing with the "Second Swipe" settings and the "Resume" setting for a specific folder. None of them made any change. Also, I am confused about the wiki page [Stop on removal with USB RFID Reader](https://github.com/MiczFlor/RPi-Jukebox-RFID/wiki/Stop-on-removal-with-USB-RFID-Reader). It seems that this hack is required for the Neuftech USB reader, but in my setup with the RC522 it is working also without any hack, right? I just got 2 seconds delay before the music stops. Is this coming from the polling interval? Or is there a interrupt settings to adjust?

**My setup:**
- Pi Zero v1.1
- HifiBerry MiniAmp
- RC522
- GPIO and RFID service enabled

Thank you for your kind help! This is a great project and we are having a lot of fun while building our first PhonieBox!

**Comment by AlvinSchiller** [comment, marked as answer]
Hey, 
i tried to reproduce your intended behavior on the current `develop` branch and it worked in my opinion, if you activate the "resume" option on the folders and also set the second swipe option to "resume playback". Can you verify this?
What i noticed though is, on switching cards the playback starts at the beginning of the song and then jumps forward to the "current" position, which sounds a bit odd (https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/878#issuecomment-672283454).

And yes the delay on removing the cards is from the polling implementation of the reader (at all 1.2 s). This cannot be adjusted (implementation detail: 0.2 seconds wait time is used for throttling + 1 second wait with a signal.alarm if no card is recognized. singal.alarm lowest accepted value is 1second ).
The hack is also just needed for the USB Reader. I use the RC522 myself and its working just fine.

**Comment by arjunsarin** [comment]
Hello @AlvinSchiller, thank your reply and taking the time looking into this!

I understand the two settings now better. So when the **"General RFID Player Control"** is set to "Placing card, ...", it will keep looking if the card is still present and (if not) stops playing. The **"Second Swipe"** option is checking if the same card is presented a second time and then executes the configured action.

However, it is still not working as desired after making the settings you suggested. I am on the `master` branch, but after `git stash; git checkout develop; git stash pop` the player did not start playing when presenting a tag. Maybe I have to re-run the installation from the develop branch..?

The folder is configured as follows:
```
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="ON"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
```

I noticed, that there is no `track1.chapters.json` in the `music` folder. Is this a related problem?
```
pi@phoniebox:~/RPi-Jukebox-RFID/shared/audiofolders/music $ ls
folder.conf  track1.mp3
```

Here is the log after placing a tag (playing 10s), removing it and then placing again (track starts from beginning) and play for 5s:
```
########### SCRIPT rfid_trigger_play.sh (2023-11-23.23:55:36) ##
Card ID '17012910860123' was used
This ID has been used before.
The shortcut points to audiofolder 'music'.
# Attempting to play: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music
# Type of play $VALUE:
$FOLDER set, not empty and dir exists: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music
  Var $LASTFOLDER:
  Var $LASTPLAYLIST:
Checking 'recursive' list? VAR $VALUE:
recursive? NO
/home/pi/RPi-Jukebox-RFID/scripts/playlist_recursive_by_folder.php --folder "music" > "/home/pi/RPi-Jukebox-RFID/playlists/music.m3u"
  Var $SECONDSWIPE: PLAY
  Var $PLAYLISTNAME: music
  Var $LASTPLAYLIST:
We must play the playlist no matter what: $PLAYPLAYLIST == yes
  VAR FOLDER: music
  VAR PLAYLISTPATH: /home/pi/RPi-Jukebox-RFID/playlists/music.m3u
########### SCRIPT playout_controls.sh (2023-11-23.23:55:37) ##
VAR COMMAND: playerstop
VAR VALUE:
current file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.mp3
chapters file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.chapters.json
chapters for extension enabled: 0
chapters for duration enabled: 0
   playerstop
#START##### SCRIPT resume_play.sh (2023-11-23.23:55:37) ##
VAR AUDIOFOLDERSPATH: /home/pi/RPi-Jukebox-RFID/shared/audiofolders
VAR COMMAND: savepos
VAR VALUE:
VAR FOLDER:
  content of /home/pi/RPi-Jukebox-RFID/shared/audiofolders//folder.conf
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="%RESUME%"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
  Now doing what COMMAND wants: savepos
VAR FOLDER from settings/Latest_Folder_Played:
  savepos FOLDER:
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="%RESUME%"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
#END####### SCRIPT resume_play.sh (2023-11-23.23:55:37) ##
remove playing.txt
########### SCRIPT playout_controls.sh (2023-11-23.23:55:37) ##
VAR COMMAND: playlistaddplay
VAR VALUE: music
current file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.mp3
chapters file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.chapters.json
chapters for extension enabled: 0
chapters for duration enabled: 0
   playlistaddplay playlist name VALUE: music
   playlistaddplay FOLDER: music
mpc load music
mpc load music && /home/pi/RPi-Jukebox-RFID/scripts/resume_play.sh -c=resume -d=music
#START##### SCRIPT resume_play.sh (2023-11-23.23:55:38) ##
VAR AUDIOFOLDERSPATH: /home/pi/RPi-Jukebox-RFID/shared/audiofolders
VAR COMMAND: resume
VAR VALUE:
VAR FOLDER: music
  content of /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/folder.conf
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="ON"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
  Now doing what COMMAND wants: resume
  savepos FOLDER: music
  entering: resume with value ON
  entering: single with value %SINGLE%
#END####### SCRIPT resume_play.sh (2023-11-23.23:55:38) ##
  echo music > /home/pi/RPi-Jukebox-RFID/scripts/../settings/Latest_Folder_Played
  VAR Latest_Folder_Played: music
  # end playout_controls.sh playlistaddplay
  Command: /home/pi/RPi-Jukebox-RFID/scripts/playout_controls.sh -c=playlistaddplay -v="music" -d="music"
########### SCRIPT playout_controls.sh (2023-11-23.23:55:50) ##
VAR COMMAND: playerpauseforce
VAR VALUE: 0.1
current file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.mp3
chapters file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.chapters.json
chapters for extension enabled: 0
chapters for duration enabled: 0
########### SCRIPT rfid_trigger_play.sh (2023-11-23.23:55:58) ##
Card ID '17012910860123' was used
This ID has been used before.
The shortcut points to audiofolder 'music'.
# Attempting to play: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music
# Type of play $VALUE:
$FOLDER set, not empty and dir exists: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music
  Var $LASTFOLDER:
  Var $LASTPLAYLIST:
Checking 'recursive' list? VAR $VALUE:
recursive? NO
/home/pi/RPi-Jukebox-RFID/scripts/playlist_recursive_by_folder.php --folder "music" > "/home/pi/RPi-Jukebox-RFID/playlists/music.m3u"
  Var $SECONDSWIPE: PLAY
  Var $PLAYLISTNAME: music
  Var $LASTPLAYLIST:
We must play the playlist no matter what: $PLAYPLAYLIST == yes
  VAR FOLDER: music
  VAR PLAYLISTPATH: /home/pi/RPi-Jukebox-RFID/playlists/music.m3u
########### SCRIPT playout_controls.sh (2023-11-23.23:55:58) ##
VAR COMMAND: playerstop
VAR VALUE:
current file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.mp3
chapters file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.chapters.json
chapters for extension enabled: 0
chapters for duration enabled: 0
   playerstop
#START##### SCRIPT resume_play.sh (2023-11-23.23:55:59) ##
VAR AUDIOFOLDERSPATH: /home/pi/RPi-Jukebox-RFID/shared/audiofolders
VAR COMMAND: savepos
VAR VALUE:
VAR FOLDER:
  content of /home/pi/RPi-Jukebox-RFID/shared/audiofolders//folder.conf
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="%RESUME%"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
  Now doing what COMMAND wants: savepos
VAR FOLDER from settings/Latest_Folder_Played:
  savepos FOLDER:
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="%RESUME%"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
#END####### SCRIPT resume_play.sh (2023-11-23.23:55:59) ##
remove playing.txt
########### SCRIPT playout_controls.sh (2023-11-23.23:55:59) ##
VAR COMMAND: playlistaddplay
VAR VALUE: music
current file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.mp3
chapters file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.chapters.json
chapters for extension enabled: 0
chapters for duration enabled: 0
   playlistaddplay playlist name VALUE: music
   playlistaddplay FOLDER: music
mpc load music
mpc load music && /home/pi/RPi-Jukebox-RFID/scripts/resume_play.sh -c=resume -d=music
#START##### SCRIPT resume_play.sh (2023-11-23.23:56:00) ##
VAR AUDIOFOLDERSPATH: /home/pi/RPi-Jukebox-RFID/shared/audiofolders
VAR COMMAND: resume
VAR VALUE:
VAR FOLDER: music
  content of /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/folder.conf
CURRENTFILENAME="%CURRENTFILENAME%"
ELAPSED="%ELAPSED%"
PLAYSTATUS="%PLAYSTATUS%"
RESUME="ON"
SHUFFLE="%SHUFFLE%"
LOOP="%LOOP%"
SINGLE="%SINGLE%"
  Now doing what COMMAND wants: resume
  savepos FOLDER: music
  entering: resume with value ON
  entering: single with value %SINGLE%
#END####### SCRIPT resume_play.sh (2023-11-23.23:56:00) ##
  echo music > /home/pi/RPi-Jukebox-RFID/scripts/../settings/Latest_Folder_Played
  VAR Latest_Folder_Played: music
  # end playout_controls.sh playlistaddplay
  Command: /home/pi/RPi-Jukebox-RFID/scripts/playout_controls.sh -c=playlistaddplay -v="music" -d="music"
########### SCRIPT playout_controls.sh (2023-11-23.23:56:06) ##
VAR COMMAND: playerpauseforce
VAR VALUE: 0.1
current file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.mp3
chapters file: /home/pi/RPi-Jukebox-RFID/shared/audiofolders/music/track1.chapters.json
chapters for extension enabled: 0
chapters for duration enabled: 0
```

**Reply by AlvinSchiller** [reply]
> I understand the two settings now better. So when the **"General RFID Player Control"** is set to "Placing card, ...", it will keep looking if the card is still present and (if not) stops playing. The **"Second Swipe"** option is checking if the same card is presented a second time and then executes the configured action.

Yes. And most importantly to unterstand is, that **"Second Swipe"** doesnt trigger if another card is used in between (thats why the "resume" on the folder is also needed).

**Reply by AlvinSchiller** [reply]
I just tested on master and have the same behavior as on develop. Very strange.
What OS Version are you using?

Im not sure if just switching the branch is supported, but as master and develop shouldn't make a difference, you could go back to master, if you haven't changed anything. Otherwise i would recommand a new installation of master or develop with a fresh image.


> CURRENTFILENAME="%CURRENTFILENAME%"
> ELAPSED="%ELAPSED%"
> PLAYSTATUS="%PLAYSTATUS%"
> RESUME="ON"
> SHUFFLE="%SHUFFLE%"
> LOOP="%LOOP%"
> SINGLE="%SINGLE%"

That doesn't look right. These %...% entries should have been replaced with the correct values.
Did you add the folder.conf manually or was it created automatically?

You could try and delete the folder.conf files and then try again.

**Reply by arjunsarin** [reply]
Good to know that master is working the same. I am on buster with the classic PhonieBox v2.4 version.

No, switching branches is not supported but switching back was restoring the initial setup.

The `folder.conf`was created by PhonieBox, not manually. But I agree, that the `%...%` look wrong. Deleting the file did not help, the same file was created, again with the place holders.

I guess next I will scrap everything and start with a new OS image and a new PhonieBox install. Let's see if this fixes it. I will come back if that solved the issue.

**Reply by AlvinSchiller** [reply]
That would be my next suggestion to start from scratch. Also i would recommend using Bullseye as Buster support is running out.

**Reply by arjunsarin** [reply]
Ok, now with a new OS (bullseye) and a fresh PhonieBox installation it works all fine. The `folder.conf` now also looks good. I don't know what was the issue before. Maybe I broke something, because at first, my username wasn't `pi` so I changed it to run all install scripts without error. Then I reinstalled PhonieBox.

@AlvinSchiller Thank you again your help and time.
